File: internal/tools/browser.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
)

func (r *Registry) registerBrowserTools() {
	r.register(ToolDef{
		Name:        "browser_navigate",
		Description: "Navigate the browser to a URL",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"session_id": map[string]any{"type": "string", "description": "Session ID"},
				"url":        map[string]any{"type": "string", "description": "URL to navigate to"},
				"wait_until": map[string]any{"type": "string", "enum": []string{"load", "domcontentloaded", "networkidle"}, "default": "load"},
			},
			"required": []string{"session_id", "url"},
		},
		Handler: r.handleBrowserNavigate,
	})

	r.register(ToolDef{
		Name:        "browser_screenshot",
		Description: "Take a screenshot of the current page (returns base64 PNG)",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"session_id": map[string]any{"type": "string", "description": "Session ID"},
				"selector":   map[string]any{"type": "string", "description": "CSS selector to screenshot (optional, defaults to full page)"},
				"full_page":  map[string]any{"type": "boolean", "description": "Capture full scrollable page", "default": false},
			},
			"required": []string{"session_id"},
		},
		Handler: r.handleBrowserScreenshot,
	})

	r.register(ToolDef{
		Name:        "browser_click",
		Description: "Click an element on the page",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"session_id": map[string]any{"type": "string", "description": "Session ID"},
				"selector":   map[string]any{"type": "string", "description": "CSS selector of element to click"},
			},
			"required": []string{"session_id", "selector"},
		},
		Handler: r.handleBrowserClick,
	})

	r.register(ToolDef{
		Name:        "browser_type",
		Description: "Type text into an input field",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"session_id": map[string]any{"type": "string", "description": "Session ID"},
				"selector":   map[string]any{"type": "string", "description": "CSS selector of input element"},
				"text":       map[string]any{"type": "string", "description": "Text to type"},
				"clear":      map[string]any{"type": "boolean", "description": "Clear field before typing", "default": false},
			},
			"required": []string{"session_id", "selector", "text"},
		},
		Handler: r.handleBrowserType,
	})

	r.register(ToolDef{
		Name:        "browser_extract",
		Description: "Extract text content from the page or a specific element",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"session_id": map[string]any{"type": "string", "description": "Session ID"},
				"selector":   map[string]any{"type": "string", "description": "CSS selector (optional, defaults to body)"},
				"max_length": map[string]any{"type": "integer", "description": "Max characters to return", "default": 5000},
			},
			"required": []string{"session_id"},
		},
		Handler: r.handleBrowserExtract,
	})

	r.register(ToolDef{
		Name:        "browser_execute",
		Description: "Execute JavaScript in the page context",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"session_id": map[string]any{"type": "string", "description": "Session ID"},
				"script":     map[string]any{"type": "string", "description": "JavaScript code to execute"},
			},
			"required": []string{"session_id", "script"},
		},
		Handler: r.handleBrowserExecute,
	})

	r.register(ToolDef{
		Name:        "browser_wait",
		Description: "Wait for an element to appear on the page",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"session_id": map[string]any{"type": "string", "description": "Session ID"},
				"selector":   map[string]any{"type": "string", "description": "CSS selector to wait for"},
				"timeout_ms": map[string]any{"type": "integer", "description": "Timeout in milliseconds", "default": 30000},
			},
			"required": []string{"session_id", "selector"},
		},
		Handler: r.handleBrowserWait,
	})

	r.register(ToolDef{
		Name:        "browser_scroll",
		Description: "Scroll the page in a given direction",
		InputSchema: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"session_id": map[string]any{"type": "string", "description": "Session ID"},
				"direction":  map[string]any{"type": "string", "enum": []string{"up", "down"}, "default": "down"},
				"amount":     map[string]any{"type": "integer", "description": "Pixels to scroll", "default": 500},
			},
			"required": []string{"session_id"},
		},
		Handler: r.handleBrowserScroll,
	})
}

func (r *Registry) handleBrowserNavigate(ctx context.Context, args map[string]any) (*ToolResult, error) {
	sessionID, _ := args["session_id"].(string)
	if sessionID == "" {
		return nil, fmt.Errorf("session_id is required")
	}

	body := map[string]any{
		"url": args["url"],
	}
	if v, ok := args["wait_until"]; ok {
		body["wait_until"] = v
	}

	resp, err := r.client.Do(ctx, "POST", "/v1/sessions/"+sessionID+"/navigate", body)
	if err != nil {
		return nil, fmt.Errorf("navigating: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	return jsonResult(result)
}

func (r *Registry) handleBrowserScreenshot(ctx context.Context, args map[string]any) (*ToolResult, error) {
	sessionID, _ := args["session_id"].(string)
	if sessionID == "" {
		return nil, fmt.Errorf("session_id is required")
	}

	body := map[string]any{}
	if v, ok := args["selector"]; ok {
		body["selector"] = v
	}
	if v, ok := args["full_page"]; ok {
		body["full_page"] = v
	}

	resp, err := r.client.Do(ctx, "POST", "/v1/sessions/"+sessionID+"/screenshot", body)
	if err != nil {
		return nil, fmt.Errorf("taking screenshot: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	// If the response contains base64 image data, return as image
	if data, ok := result["data"].(string); ok {
		return imageResult(data, "image/png"), nil
	}

	return jsonResult(result)
}

func (r *Registry) handleBrowserClick(ctx context.Context, args map[string]any) (*ToolResult, error) {
	sessionID, _ := args["session_id"].(string)
	if sessionID == "" {
		return nil, fmt.Errorf("session_id is required")
	}

	body := map[string]any{
		"selector": args["selector"],
	}

	resp, err := r.client.Do(ctx, "POST", "/v1/sessions/"+sessionID+"/click", body)
	if err != nil {
		return nil, fmt.Errorf("clicking: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	return jsonResult(result)
}

func (r *Registry) handleBrowserType(ctx context.Context, args map[string]any) (*ToolResult, error) {
	sessionID, _ := args["session_id"].(string)
	if sessionID == "" {
		return nil, fmt.Errorf("session_id is required")
	}

	body := map[string]any{
		"selector": args["selector"],
		"text":     args["text"],
	}
	if v, ok := args["clear"]; ok {
		body["clear"] = v
	}

	resp, err := r.client.Do(ctx, "POST", "/v1/sessions/"+sessionID+"/type", body)
	if err != nil {
		return nil, fmt.Errorf("typing: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	return jsonResult(result)
}

func (r *Registry) handleBrowserExtract(ctx context.Context, args map[string]any) (*ToolResult, error) {
	sessionID, _ := args["session_id"].(string)
	if sessionID == "" {
		return nil, fmt.Errorf("session_id is required")
	}

	body := map[string]any{}
	if v, ok := args["selector"]; ok {
		body["selector"] = v
	}
	if v, ok := args["max_length"]; ok {
		body["max_length"] = v
	}

	resp, err := r.client.Do(ctx, "POST", "/v1/sessions/"+sessionID+"/extract", body)
	if err != nil {
		return nil, fmt.Errorf("extracting: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	return jsonResult(result)
}

func (r *Registry) handleBrowserExecute(ctx context.Context, args map[string]any) (*ToolResult, error) {
	sessionID, _ := args["session_id"].(string)
	if sessionID == "" {
		return nil, fmt.Errorf("session_id is required")
	}

	body := map[string]any{
		"script": args["script"],
	}

	resp, err := r.client.Do(ctx, "POST", "/v1/sessions/"+sessionID+"/execute", body)
	if err != nil {
		return nil, fmt.Errorf("executing: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	return jsonResult(result)
}

func (r *Registry) handleBrowserWait(ctx context.Context, args map[string]any) (*ToolResult, error) {
	sessionID, _ := args["session_id"].(string)
	if sessionID == "" {
		return nil, fmt.Errorf("session_id is required")
	}

	body := map[string]any{
		"selector": args["selector"],
	}
	if v, ok := args["timeout_ms"]; ok {
		body["timeout_ms"] = v
	}

	resp, err := r.client.Do(ctx, "POST", "/v1/sessions/"+sessionID+"/wait", body)
	if err != nil {
		return nil, fmt.Errorf("waiting: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	return jsonResult(result)
}

func (r *Registry) handleBrowserScroll(ctx context.Context, args map[string]any) (*ToolResult, error) {
	sessionID, _ := args["session_id"].(string)
	if sessionID == "" {
		return nil, fmt.Errorf("session_id is required")
	}

	body := map[string]any{}
	if v, ok := args["direction"]; ok {
		body["direction"] = v
	}
	if v, ok := args["amount"]; ok {
		body["amount"] = v
	}

	resp, err := r.client.Do(ctx, "POST", "/v1/sessions/"+sessionID+"/scroll", body)
	if err != nil {
		return nil, fmt.Errorf("scrolling: %w", err)
	}

	var result map[string]any
	if err := json.Unmarshal(resp, &result); err != nil {
		return nil, fmt.Errorf("parsing response: %w", err)
	}

	return jsonResult(result)
}

File: src/tools/browser.ts
import type { Tool } from "@modelcontextprotocol/sdk/types.js";
import type { MeshbrowClient } from "../client.js";

export const browserTools: Tool[] = [
  {
    name: "browser_navigate",
    description:
      "Navigate the browser to a URL. Waits for the page to load before returning.",
    inputSchema: {
      type: "object" as const,
      properties: {
        session_id: {
          type: "string",
          description: "The session ID of the browser to navigate",
        },
        url: {
          type: "string",
          description: "The URL to navigate to",
        },
        wait: {
          type: "string",
          enum: ["load", "domcontentloaded", "networkidle"],
          description: "Wait condition before returning (default: load)",
        },
        timeout: {
          type: "number",
          description: "Navigation timeout in seconds (default: 30)",
        },
      },
      required: ["session_id", "url"],
    },
  },
  {
    name: "browser_screenshot",
    description:
      "Take a screenshot of the current page. Returns a base64-encoded PNG image.",
    inputSchema: {
      type: "object" as const,
      properties: {
        session_id: {
          type: "string",
          description: "The session ID to screenshot",
        },
        full_page: {
          type: "boolean",
          description: "Capture the full scrollable page (default: false)",
        },
      },
      required: ["session_id"],
    },
  },
  {
    name: "browser_click",
    description: "Click an element on the page by CSS selector.",
    inputSchema: {
      type: "object" as const,
      properties: {
        session_id: {
          type: "string",
          description: "The session ID",
        },
        selector: {
          type: "string",
          description: "CSS selector of the element to click",
        },
      },
      required: ["session_id", "selector"],
    },
  },
  {
    name: "browser_type",
    description: "Type text into a focused element or a specified input field.",
    inputSchema: {
      type: "object" as const,
      properties: {
        session_id: {
          type: "string",
          description: "The session ID",
        },
        selector: {
          type: "string",
          description: "CSS selector of the input element",
        },
        text: {
          type: "string",
          description: "Text to type into the element",
        },
        clear: {
          type: "boolean",
          description: "Clear the field before typing (default: false)",
        },
      },
      required: ["session_id", "selector", "text"],
    },
  },
  {
    name: "browser_extract",
    description:
      "Extract text content or data from the page using a CSS selector or JavaScript expression.",
    inputSchema: {
      type: "object" as const,
      properties: {
        session_id: {
          type: "string",
          description: "The session ID",
        },
        selector: {
          type: "string",
          description: "CSS selector to extract text from (returns innerText of all matches)",
        },
        script: {
          type: "string",
          description:
            "JavaScript expression to evaluate and return (alternative to selector)",
        },
      },
      required: ["session_id"],
    },
  },
  {
    name: "browser_execute",
    description: "Execute arbitrary JavaScript in the browser and return the result.",
    inputSchema: {
      type: "object" as const,
      properties: {
        session_id: {
          type: "string",
          description: "The session ID",
        },
        script: {
          type: "string",
          description: "JavaScript code to execute in the browser context",
        },
      },
      required: ["session_id", "script"],
    },
  },
  {
    name: "browser_wait",
    description: "Wait for an element to appear on the page, or wait a specified duration.",
    inputSchema: {
      type: "object" as const,
      properties: {
        session_id: {
          type: "string",
          description: "The session ID",
        },
        selector: {
          type: "string",
          description: "CSS selector to wait for",
        },
        timeout: {
          type: "number",
          description: "Maximum time to wait in seconds (default: 30)",
        },
      },
      required: ["session_id"],
    },
  },
  {
    name: "browser_scroll",
    description: "Scroll the page in a direction or to a specific element.",
    inputSchema: {
      type: "object" as const,
      properties: {
        session_id: {
          type: "string",
          description: "The session ID",
        },
        direction: {
          type: "string",
          enum: ["up", "down"],
          description: "Scroll direction (default: down)",
        },
        selector: {
          type: "string",
          description: "CSS selector to scroll into view (overrides direction)",
        },
        amount: {
          type: "number",
          description: "Pixels to scroll (default: 500)",
        },
      },
      required: ["session_id"],
    },
  },
];

export async function handleBrowserTool(
  client: MeshbrowClient,
  name: string,
  params: Record<string, unknown>
): Promise<unknown> {
  const sessionId = params.session_id as string;

  switch (name) {
    case "browser_navigate":
      return client.post(`/v1/sessions/${sessionId}/navigate`, {
        url: params.url,
        wait: params.wait || "load",
        timeout: params.timeout || 30,
      });

    case "browser_screenshot":
      return client.post(`/v1/sessions/${sessionId}/screenshot`, {
        full_page: params.full_page || false,
      });

    case "browser_click":
      return client.post(`/v1/sessions/${sessionId}/execute`, {
        script: `document.querySelector('${params.selector}').click()`,
      });

    case "browser_type": {
      const clear = params.clear ? `document.querySelector('${params.selector}').value = '';` : "";
      return client.post(`/v1/sessions/${sessionId}/execute`, {
        script: `${clear}
          const el = document.querySelector('${params.selector}');
          el.focus();
          el.value = '${(params.text as string).replace(/'/g, "\\'")}';
          el.dispatchEvent(new Event('input', { bubbles: true }));
          el.dispatchEvent(new Event('change', { bubbles: true }));`,
      });
    }

    case "browser_extract": {
      let script: string;
      if (params.script) {
        script = params.script as string;
      } else if (params.selector) {
        script = `Array.from(document.querySelectorAll('${params.selector}')).map(el => el.innerText).join('\\n')`;
      } else {
        script = "document.body.innerText";
      }
      return client.post(`/v1/sessions/${sessionId}/execute`, { script });
    }

    case "browser_execute":
      return client.post(`/v1/sessions/${sessionId}/execute`, {
        script: params.script,
      });

    case "browser_wait": {
      if (params.selector) {
        const timeout = ((params.timeout as number) || 30) * 1000;
        return client.post(`/v1/sessions/${sessionId}/execute`, {
          script: `await new Promise((resolve, reject) => {
            const el = document.querySelector('${params.selector}');
            if (el) return resolve(true);
            const observer = new MutationObserver(() => {
              if (document.querySelector('${params.selector}')) {
                observer.disconnect();
                resolve(true);
              }
            });
            observer.observe(document.body, { childList: true, subtree: true });
            setTimeout(() => { observer.disconnect(); reject(new Error('Timeout')); }, ${timeout});
          })`,
        });
      }
      return { status: "ok" };
    }

    case "browser_scroll": {
      let script: string;
      if (params.selector) {
        script = `document.querySelector('${params.selector}').scrollIntoView({ behavior: 'smooth' })`;
      } else {
        const amount = (params.amount as number) || 500;
        const direction = params.direction === "up" ? -amount : amount;
        script = `window.scrollBy(0, ${direction})`;
      }
      return client.post(`/v1/sessions/${sessionId}/execute`, { script });
    }

    default:
      throw new Error(`Unknown browser tool: ${name}`);
  }
}
